Add pyvista 3 d visualization

MaHaWo opened the pull request:
- adds the pyvista plotting utilities used in the toy-models notebooks in a generalized way. 
  - adds 2 functions: `facet_grid_3D` and `facet_grid_animate_3D`  which work much like [seaborn facetgrid](https://seaborn.pydata.org/generated/seaborn.FacetGrid.html), but as a function instead of a class. 
  - add some utilities to make sme output work well with pyvista and for plot layouting
  - add tests
 - encountered problems with Python3.7, and everything below 3.9 is [beyond its end of life anyway](https://devguide.python.org/versions/), therefore I updated the python version tags in the `pyproject.toml` and the test gh workflow to support python3.9 to 3.13

## Problem: 
- sonar action is not working b/c of missing secret(?) and is deprecated and should be updated. Should that be done here?
- pyvista depends on vtk and sme bundles vtk (?), which clashes and creates eternally hanging exectution, crashing notebook kernels or hanging execution. Hence I added a rather hacky "initialization code" that's guarded by an environment flag. Not pretty and probably not stable over time, but I am not sure how to fix this properly (integrate pyvista into sme perhaps? )

MaHaWo committed:
add comments and data

MaHaWo committed:
improve pyvista plotting functions

MaHaWo committed:
add tests for pyvista plotting utils

MaHaWo committed:
finish utils test

MaHaWo committed:
reorganize tests

MaHaWo committed:
make test compatible with coverage reporting

MaHaWo committed:
remove superfluous function, make docstring better

MaHaWo committed:
adjust python versions

MaHaWo committed:
add pyvista headless display to

MaHaWo committed:
make code compatible with old python versions

MaHaWo committed:
try to fix docstring issues

MaHaWo committed:
fix documentation issue

MaHaWo committed:
fix issues with default args

MaHaWo committed:
fix bug in constructor call

lkeegan reviewed (commented):
thanks, looks very nice!

it would be great if you could also add an example of using this functionality with sme to the plot notebook (which currently shows how to plot results from a 2d model, it would be nice to have the same for a 3d model with your new plotting functions):

- notebook: https://github.com/spatial-model-editor/sme-contrib/blob/main/docs/notebooks/plot.ipynb
- generated docs: https://sme-contrib.readthedocs.io/en/latest/notebooks/plot.html

and a related point: it's not immediately clear to me as a user how I should use these functions to plot the simulation results - would it be possible to adjust the API to take a `sme.SimulationResult` similar to the existing 2d plotting functions? (see e.g. 
https://sme-contrib.readthedocs.io/en/latest/_autosummary/sme_contrib.plot.concentration_heatmap.html)

MaHaWo committed:
add functions that eat sme.SimulationResult

MaHaWo committed:
add 3D wrapper functions and docs notebook stuff

MaHaWo committed:
correct type annotations

MaHaWo committed:
correct typos

MaHaWo committed:
make notebook work

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaHaWo committed:
try to add static backend to make pv work

MaHaWo committed:
Merge branch 'add-pyvista-3D-visualization' of github.com:spatial-model-editor/sme-contrib into add-pyvista-3D-visualization

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaHaWo committed:
try to make docs thing work again damnit

MaHaWo committed:
Merge branch 'add-pyvista-3D-visualization' of github.com:spatial-model-editor/sme-contrib into add-pyvista-3D-visualization

MaHaWo committed:
fix import issue

MaHaWo committed:
fix improve README section in notebook

MaHaWo committed:
use client mode for pv to try and make pipeline work

MaHaWo committed:
try to make doc build on ci work

MaHaWo committed:
try once more to make docs ci work

MaHaWo commented:
ok, I added wrappers that make it easier to plot concentration images and added some (rudimentary) examples to the notebook along with some descriptive text. The wrapper code should make it clearer how to use the `facetgrid`  functions. I'm currently still trying to get the docs to build on the pipeline...

MaHaWo committed:
adjust headings

MaHaWo committed:
add hacky fix for sme<->pyvista vtk incompatability

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaHaWo committed:
add guard flag to prevent init of vtk

MaHaWo committed:
add env flag to make vtk init configurable

MaHaWo committed:
rename init guard flag to something a little simpler, add readme

MaHaWo committed:
add some more context to the 3d plot notebook

MaHaWo committed:
reset optimize notebook

MaHaWo committed:
make failure-path a separate test for 3d facet plot

MaHaWo commented:
ok, I added examples to the notebook, and some wrapper functions that plot concentrations in 3d. The example is a little bit silly b/c it essentially uses a 2D simulation to make a 3D plot for exec time reasons.

MaHaWo commented:
I am not entirely sure why sonarcloud/cube/whatever can't find its token anymore... I'm normally not using these services.

lkeegan committed:
remove sonar

lkeegan committed:
bump actions and supported python versions

lkeegan committed:
update pre-commit

lkeegan committed:
fix off-by-one indexing of titles

lkeegan committed:
update docstring/typing for plotfunc_kwargs

lkeegan committed:
clear any existing actors at the start of create_frame

lkeegan committed:
skip any empty layout cells (e.g. 5 -> (2, 3) layout with last cell empty)

lkeegan reviewed (approved):
thanks, looks very nice!

a few minor suggestions:

- it would be nice if the 2d and 3d plotting functions had more consistent names (we can also rename the existing ones)
<img width="1577" height="575" alt="Image" src="https://github.com/user-attachments/assets/fa970af2-5de4-4343-a992-e886a01485b0" />

- in the example notebook we could first show the high-level functions, then have the lower level interface as an advanced section afterwards: always nice when the first thing the user sees is a one-liner that does what they wanted to do :-)
- the env var / order of module loading stuff can (hopefully) be removed now that the issue has been fixed upstream in sme

MaHaWo commented:
- [x] align names
- [x] reorder plotting notebook 
- [x] remove weird module loading workaround

MaHaWo committed:
align naming of functions

MaHaWo committed:
reorder notebook cells

MaHaWo committed:
correct headings in plot notebook

MaHaWo committed:
adjust tests

MaHaWo committed:
remove pyvista import workaround

MaHaWo committed:
remove some superfluous imports

MaHaWo committed:
rename pyvista colormap function and remove lazy pyvista loading
